File: server/services/extractor.js
/**
 * Multi-format text extractor.
 * Supports: PDF, DOCX, XLSX/XLS, PNG/JPG (via Tesseract OCR)
 */

import fs from 'fs'
import path from 'path'
import pdfParse from 'pdf-parse/lib/pdf-parse.js'
import mammoth from 'mammoth'
import ExcelJS from 'exceljs'
import Tesseract from 'tesseract.js'
import { GoogleGenerativeAI } from '@google/generative-ai'

/**
 * Extract text pages from a file.
 * Returns: { text, pageNumber, documentName }[]
 */
export async function extractText(filePath) {
  const ext = path.extname(filePath).toLowerCase()
  const documentName = path.basename(filePath)

  switch (ext) {
    case '.pdf':   return extractPDF(filePath, documentName)
    case '.docx':  return extractDOCX(filePath, documentName)
    case '.xlsx':
    case '.xls':   return await extractExcel(filePath, documentName)
    case '.png':
    case '.jpg':
    case '.jpeg':  return extractImage(filePath, documentName)
    default:
      throw new Error(`Unsupported file type: ${ext}`)
  }
}

// ── PDF ──────────────────────────────────────────────────────────────────────
async function extractPDF(filePath, documentName) {
  const buffer = fs.readFileSync(filePath)
  const data = await pdfParse(buffer)

  const text = data.text?.trim() || ''

  // If too little text, try Gemini Vision OCR on the PDF
  if (text.length < 100) {
    return extractPDFWithVision(filePath, documentName)
  }

  // Split into pseudo-pages by form-feed or chunk by paragraph
  const pages = text.split(/\f/).filter((p) => p.trim().length > 0)
  if (pages.length === 0) return []

  return pages.map((pageText, i) => ({
    text: pageText.trim(),
    pageNumber: i + 1,
    documentName,
  }))
}

async function extractPDFWithVision(filePath, documentName) {
  // Use Gemini Vision to OCR the PDF as a whole file
  const genAI = new GoogleGenerativeAI(process.env.GEMINI_API_KEY)
  const model = genAI.getGenerativeModel({ model: 'gemini-2.5-flash' })

  const fileBuffer = fs.readFileSync(filePath)
  const base64 = fileBuffer.toString('base64')

  const result = await model.generateContent([
    {
      inlineData: {
        mimeType: 'application/pdf',
        data: base64,
      },
    },
    'Extract all text from this PDF exactly as it appears. Preserve paragraph and page structure. Return only the extracted text.',
  ])

  const text = result.response.text().trim()
  if (!text) return []

  return [{ text, pageNumber: 1, documentName }]
}

// ── DOCX ─────────────────────────────────────────────────────────────────────
async function extractDOCX(filePath, documentName) {
  const result = await mammoth.extractRawText({ path: filePath })
  const text = result.value?.trim()
  if (!text) return []
  return [{ text, pageNumber: 1, documentName }]
}

// ── Excel ────────────────────────────────────────────────────────────────────
async function extractExcel(filePath, documentName) {
  const workbook = new ExcelJS.Workbook()
  await workbook.xlsx.readFile(filePath)
  const pages = []

  workbook.eachSheet((sheet, i) => {
    const rows = []
    sheet.eachRow((row) => {
      const cells = row.values.slice(1).map((v) =>
        v === null || v === undefined ? '' : String(v?.result ?? v)
      )
      rows.push(cells.join('\t'))
    })
    const text = `Sheet: ${sheet.name}\n\n${rows.join('\n')}`.trim()
    if (text.length > 10) {
      pages.push({ text, pageNumber: i, documentName })
    }
  })

  return pages
}

// ── Images (Tesseract OCR) ───────────────────────────────────────────────────
async function extractImage(filePath, documentName) {
  const { data: { text } } = await Tesseract.recognize(filePath, 'eng', {
    logger: () => {},
  })
  if (!text?.trim()) return []
  return [{ text: text.trim(), pageNumber: 1, documentName }]
}

File: backend/services/extractor.py
"""
Server-side extractor — kept for the Python backend only.
In normal operation the Node.js server accepts client-side chunks, so this
file is only used if you run the FastAPI backend directly.
"""

import os
import io
from pathlib import Path
from typing import List, Dict

from dotenv import load_dotenv
import fitz  # PyMuPDF
from docx import Document
import pandas as pd
from PIL import Image
from google import genai

VISION_MODEL = "gemini-2.5-flash"
_ENV_PATH = Path(__file__).parent.parent / ".env"


def _key() -> str:
    load_dotenv(dotenv_path=_ENV_PATH, override=True)
    return os.environ["GEMINI_API_KEY"]


def _ocr_image(img: Image.Image) -> str:
    buf = io.BytesIO()
    img.save(buf, format="PNG")
    image_bytes = buf.getvalue()
    with genai.Client(api_key=_key()) as client:
        response = client.models.generate_content(
            model=VISION_MODEL,
            contents=[
                {"mime_type": "image/png", "data": image_bytes},
                "Extract all text from this image exactly as it appears. "
                "Preserve paragraph structure. Return only the extracted text.",
            ],
        )
    return response.text.strip() if response.text else ""


def extract_text_from_pdf(file_path: str) -> List[Dict]:
    pages = []
    doc = fitz.open(file_path)
    doc_name = os.path.basename(file_path)
    for page_num, page in enumerate(doc, start=1):
        text = page.get_text().strip()
        if text:
            pages.append({"text": text, "page_number": page_num, "document_name": doc_name})
    doc.close()
    return pages


def extract_scanned_pdf(file_path: str) -> List[Dict]:
    pages = []
    doc = fitz.open(file_path)
    doc_name = os.path.basename(file_path)
    for page_num, page in enumerate(doc, start=1):
        mat = fitz.Matrix(2, 2)
        pix = page.get_pixmap(matrix=mat)
        img = Image.open(io.BytesIO(pix.tobytes("png")))
        text = _ocr_image(img)
        if text:
            pages.append({"text": text, "page_number": page_num, "document_name": doc_name})
    doc.close()
    return pages


def extract_text_from_docx(file_path: str) -> List[Dict]:
    doc = Document(file_path)
    parts: List[str] = []

    for para in doc.paragraphs:
        if para.text.strip():
            parts.append(para.text.strip())

    # Include table cell text (previously missing)
    for table in doc.tables:
        for row in table.rows:
            for cell in row.cells:
                if cell.text.strip():
                    parts.append(cell.text.strip())

    text = "\n".join(parts)
    if not text:
        return []
    return [{"text": text, "page_number": 1, "document_name": os.path.basename(file_path)}]


def extract_text_from_excel(file_path: str) -> List[Dict]:
    # Use sheet index as page_number (1-based) so multi-sheet docs cite correctly
    sheets = pd.read_excel(file_path, sheet_name=None)
    pages = []
    doc_name = os.path.basename(file_path)
    for sheet_index, (sheet_name, df) in enumerate(sheets.items(), start=1):
        text = f"Sheet: {sheet_name}\n\n{df.to_string(index=False)}"
        pages.append({"text": text, "page_number": sheet_index, "document_name": doc_name})
    return pages


def extract_text_from_image(file_path: str) -> List[Dict]:
    img = Image.open(file_path)
    text = _ocr_image(img)
    return [{"text": text, "page_number": 1, "document_name": os.path.basename(file_path)}]


def extract(file_path: str) -> List[Dict]:
    ext = os.path.splitext(file_path)[1].lower()

    if ext == ".pdf":
        pages = extract_text_from_pdf(file_path)
        if len(" ".join(p["text"] for p in pages).strip()) < 100:
            pages = extract_scanned_pdf(file_path)
        return pages

    if ext == ".docx":
        return extract_text_from_docx(file_path)

    if ext in {".xlsx", ".xls"}:
        return extract_text_from_excel(file_path)

    if ext in {".png", ".jpg", ".jpeg"}:
        return extract_text_from_image(file_path)

    raise ValueError(f"Unsupported file type: {ext}")
